Tools/AP_Periph/can.cpp

```diff
@@ -2356,6 +2359,192 @@ void AP_Periph_FW::can_send_ADSB(struct __mavlink_adsb_vehicle_t &msg)
+                UAVCAN_EQUIPMENT_ICE_RECIPROCATING_STATUS_FLAG_FUEL_PRESSURE_BELOW_NOMINAL;
+            break;
+        case Fuel_Pressure_Status::ABOVE_NOMINAL:
+            pkt.flags |=
+                UAVCAN_EQUIPMENT_ICE_RECIPROCATING_STATUS_FLAG_FUEL_PRESSURE_SUPPORTED |
+                UAVCAN_EQUIPMENT_ICE_RECIPROCATING_STATUS_FLAG_FUEL_PRESSURE_ABOVE_NOMINAL;
+            break;
+        }
+
+        switch (state.oil_pressure_status) {
+        case Oil_Pressure_Status::NOT_SUPPORTED:
+            break;
+        case Oil_Pressure_Status::OK:
+            pkt.flags |= UAVCAN_EQUIPMENT_ICE_RECIPROCATING_STATUS_FLAG_OIL_PRESSURE_SUPPORTED;
+            break;
+        case Oil_Pressure_Status::BELOW_NOMINAL:
+            pkt.flags |=
+                UAVCAN_EQUIPMENT_ICE_RECIPROCATING_STATUS_FLAG_OIL_PRESSURE_SUPPORTED |
+                UAVCAN_EQUIPMENT_ICE_RECIPROCATING_STATUS_FLAG_OIL_PRESSURE_BELOW_NOMINAL;
+            break;
+        case Oil_Pressure_Status::ABOVE_NOMINAL:
+            pkt.flags |=
+                UAVCAN_EQUIPMENT_ICE_RECIPROCATING_STATUS_FLAG_OIL_PRESSURE_SUPPORTED |
+                UAVCAN_EQUIPMENT_ICE_RECIPROCATING_STATUS_FLAG_OIL_PRESSURE_ABOVE_NOMINAL;
+            break;
+        }
+
+        switch (state.detonation_status) {
+        case Detonation_Status::NOT_SUPPORTED:
+            break;
+        case Detonation_Status::NOT_OBSERVED:
+            pkt.flags |=
+                UAVCAN_EQUIPMENT_ICE_RECIPROCATING_STATUS_FLAG_DETONATION_SUPPORTED;
+            break;
+        case Detonation_Status::OBSERVED:
+            pkt.flags |=
+                UAVCAN_EQUIPMENT_ICE_RECIPROCATING_STATUS_FLAG_DETONATION_SUPPORTED |
+                UAVCAN_EQUIPMENT_ICE_RECIPROCATING_STATUS_FLAG_DETONATION_OBSERVED;
+            break;
+        }
+
+        switch (state.misfire_status) {
+        case Misfire_Status::NOT_SUPPORTED:
+            break;
+        case Misfire_Status::NOT_OBSERVED:
+            pkt.flags |=
+                UAVCAN_EQUIPMENT_ICE_RECIPROCATING_STATUS_FLAG_MISFIRE_SUPPORTED;
+            break;
+        case Misfire_Status::OBSERVED:
+            pkt.flags |=
+                UAVCAN_EQUIPMENT_ICE_RECIPROCATING_STATUS_FLAG_MISFIRE_SUPPORTED |
+                UAVCAN_EQUIPMENT_ICE_RECIPROCATING_STATUS_FLAG_MISFIRE_OBSERVED;
+            break;
+        }
+
+        switch (state.debris_status) {
+        case Debris_Status::NOT_SUPPORTED:
+            break;
+        case Debris_Status::NOT_DETECTED:
+            pkt.flags |=
+                UAVCAN_EQUIPMENT_ICE_RECIPROCATING_STATUS_FLAG_DEBRIS_SUPPORTED;
+            break;
+        case Debris_Status::DETECTED:
+            pkt.flags |=
+                UAVCAN_EQUIPMENT_ICE_RECIPROCATING_STATUS_FLAG_DEBRIS_SUPPORTED |
+                UAVCAN_EQUIPMENT_ICE_RECIPROCATING_STATUS_FLAG_DEBRIS_DETECTED;
+            break;
+        }
+
+        pkt.engine_load_percent = state.engine_load_percent;
```

Comment:
Could have created this packet entirely within the declaration (inside the {}).  Helps to catch where you're narrowing your data....

Reply:
harder to debug with gdb